golem-worker-service-base/src/evaluator/component_elements.rs

```diff
@@ -0,0 +1,218 @@
+    component_metadata_service: Arc<dyn ComponentMetadataService + Sync + Send>,
+    component_elements_cache:
+        Cache<(ComponentId, ComponentVersion), (), ComponentElements, MetadataFetchError>,
+    currently_running_version_cache: Cache<WorkerId, (), ComponentVersion, MetadataFetchError>,
+}
+
+impl DefaultComponentElementsFetch {
+    pub(crate) fn new(
+        metadata_fetcher: Arc<dyn ComponentMetadataService + Sync + Send>,
+        max_cache_size: usize,
+    ) -> Self {
+        DefaultComponentElementsFetch {
+            component_metadata_service: metadata_fetcher,
+            component_elements_cache: Cache::new(
+                Some(max_cache_size),
+                golem_common::cache::FullCacheEvictionMode::LeastRecentlyUsed(1),
+                BackgroundEvictionMode::None,
+                "worker_gateway_component_info",
+            ),
+            currently_running_version_cache: Cache::new(
+                Some(max_cache_size),
+                golem_common::cache::FullCacheEvictionMode::LeastRecentlyUsed(1),
+                BackgroundEvictionMode::None,
+                "worker_gateway_running_version_info",
+            ),
+        }
+    }
+
+    pub(crate) async fn get_currently_running_version_from_cache(
+        &self,
+        worker_id: &WorkerId,
+    ) -> Result<ComponentVersion, MetadataFetchError> {
+        self.currently_running_version_cache
+            .get_or_insert_simple(&worker_id.clone(), || {
+                let component_metadata_service = self.component_metadata_service.clone();
+                let worker_id = worker_id.clone();
+                Box::pin(async move {
+                    component_metadata_service
+                        .get_active_component_in_worker(&worker_id)
+                        .await
+                })
+            })
+            .await
+    }
+
+    pub(crate) async fn get_component_elements_from_cache(
+        &self,
+        component_id: &ComponentId,
+        version: ComponentVersion,
+    ) -> Result<ComponentElements, MetadataFetchError> {
+        self.component_elements_cache
+            .get_or_insert_simple(&(component_id.clone(), version), || {
+                let component_metadata_service = self.component_metadata_service.clone();
+                let component_id = component_id.clone();
+
+                Box::pin(async move {
+                    let component = component_metadata_service
+                        .get_component_metadata(&component_id, version)
+                        .await?;
+                    Ok(ComponentElements::from_component_metadata(
+                        component.metadata,
+                    ))
+                })
+            })
+            .await
+    }
+
+    pub(crate) async fn get_component_elements_from_cache_latest_version(
+        &self,
+        worker_id: &WorkerId,
+    ) -> Result<ComponentElements, MetadataFetchError> {
+        let latest_component_version_details = self
```

Comment:
Note that, previous to this approach, we were simply forwarding everything to worker executor. That is Rib evaluation context hardly knew anything about this.

Having no symbol-table info for evaluating Rib (meaning we simply forward everything to worker-executor) is not something I am a bit hesitant to do. Even to forward everything to worker-executor and not type checking is a an approach that's always _after the fact_, meaning, the rib evaluator can still (possibly) make mistakes because it couldn't decide things properly on what to do due to missing type informations and worker-executor will always return error. I think forwarding things as such is also hacky approach, if we think about any future improvements for Rib. If the boundary of Rib is well set (ex: `foo(1)` is always a function call and not constructing a variant), then may be forwarding everything to worker may make some more sense though.

Open to discussions there, but it will be more of _reverting to where we were before_ , than _changing_ this approach.

Reply:
It's unfortunate if it would be a (partial) revert but I feel like we did not properly think through every aspect of this and that's why we are in this current state. We can continuously learn and reevaluate or idea of what we need to do.

I don't agree that leaving the validation to the worker executor is "hacky" - that's a perfectly valid validation point and done where every necessary information must be available anyway to run the worker.

Could you, to help the discussion, collect a more concrete list of what do we need the above mentioned symbol table info for - and what do we sacrifice if we would not have it? It is not clear to me. Is it just because of the "worker invocation looks like an inline function call" feature that I remember we were arguing about, or something else?

Reply:
Also if we will support the new dynamic dispatch feature, we won't have any metadata for those components.

The "worker proxy" part of worker service shouldn't require anything else than the cached routing table for forwarding invocations. Right now it requires type information but only for the json invocation api because it's ambiguous and cannot be interpreted without knowing the types. But that interpreter can be moved one layer down to the executor. 

The worker bridge part requires the metadata for evaluating Rib and I more and more think that it's a mistake

Reply:
>The worker bridge part requires the metadata for evaluating Rib and I more and more think that it's a mistake

Is this a mistake? Or it is becoming difficult due to dynamic dispatch feature?

Reply:
I think it is a mistake because it requires a very complex (or even impossible) caching logic and/or performance penalty, as I tried to pointing out.

Reply:
I do feel like there is some place where validation must be possible, even though it doesn't have to be Worker Gateway. The reason is tooling and DX: when a user is writing a Rib script, auto-complete should be possible, and type errors, to the extent they can be known in advance, should be eagerly reported. However, these could occur in some service that is separate from Worker Gateway, and which exists to do pre-compilation, typing, validation, auto-complete, etc.

Reply:
If we would have such a feature it would be tied to a specific component version and not a specific worker. So I could imagine an explicit user triggered validation that contains the targeted component version as a parameter and would be triggered during deployments or live editing in the UI.

We could also permanently associate a component version with a api definition but that would lead to the same problem as we don't have per worker information (what version they use) on the worker service level